36 タグ一覧取得APIの作成

Nox8931 opened the pull request:
## 関連Issue
<!-- #の後ろに対応するissue番号を書く -->
Closes #36 

## やったこと
タグ一覧取得APIの作成

## 動作確認
<!-- テスト内容など -->

- [ ] SwagerUIでテストが完了した


<img width="904" height="893" alt="一覧取得テスト" src="https://github.com/user-attachments/assets/b58497d7-4170-4edf-8a9e-84bdda6211f8" />
## 参考にした資料

Nox8931 committed:
feat: リポジトリ層にタグ取得メソッドを追加

- VoiceMemoRepositoryにfindTagsWithCountByUserIdメソッドを追加
- MemoJpaDataRepositoryにネイティブクエリを追加
- VoiceMemoRepositoryImplに実装を追加

Nox8931 committed:
feat:タグ一覧取得ユースケースを実装

Nox8931 committed:
feat:プレゼンテーション層タグ取得APIエンドポイント追加

junhat6 commented:
次回からでいいの、コミットメッセージのお尻にissue番号を記載して欲しいです。
コミットメッセージ #36 みたいな感じで
これは後で、コミット履歴を見た時にどのissue番号の時のコミットなのかをわかりやすく見るためです。
添付した画像が例
<img width="870" height="528" alt="image" src="https://github.com/user-attachments/assets/def8e54e-66fc-421a-9638-4bd3e4a7bb2f" />
#22みたいなissue番号は勝手にリンク化もされるので、後で対応しているissueが参照しやすい

junhat6 commented on MANUAL_TESTING.md:
ちょっと詳しく書きすぎです。リポジトリルートの.mdファイルはcluadeなどのllm cliが読み取る可能性があるので、基本的にはドキュメント量は少なくして欲しいです。
人間が読むにしてもテキスト量が多いと思います。
また、このドキュメント自体も必要かどうかがあって、長期的に必要かどうかでドキュメントとしてリポジトリに置くかどうかを判断して欲しいです。

Nox8931 replied:
削除対応しました

junhat6 commented on src/main/kotlin/com/assari/voicebooklm/infrastructure/postgres_jpa/memo/MemoJpaDataRepository.kt:
申し訳なんですが、ファイルの命名をMemoJpaDataRepositoryから
MemoJpaRepositoryに変更してもらっていいですか？
Dataって指す内容が大きすぎて不必要なんで変えといて欲しいです。

junhat6 commented on src/main/kotlin/com/assari/voicebooklm/infrastructure/postgres_jpa/memo/VoiceMemoRepositoryImpl.kt:
```diff
@@ -28,4 +28,14 @@ class VoiceMemoRepositoryImpl(
+            // ネイティブクエリの結果: [tag名(String), count(Long)]
+            val tagName = row[0] as String
+            val count = (row[1] as Number).toInt()
+            Pair(tagName, count)
```
インデックスで取ってきてるのがすごく怖いというか不安です。この辺に関しては違う実装方針がありそうです。
　copilot君がレビューしてくれている内容はインデックスに他から直接アクセスできないようにするみたいな変更の方針ですが、他の方針考えてください。インデックスはなるべく使わないで欲しいです。

junhat6 commented on src/main/kotlin/com/assari/voicebooklm/presentation/controller/tag/TagController.kt:
```diff
@@ -0,0 +1,66 @@
+@RequestMapping("/api/tags")
+@Tag(name = "Tags", description = "タグ関連 API")
+class TagController(
+    private val voiceMemoRepository: VoiceMemoRepository,
```
今、コントローラ層依存関係ボコボコなんですけど、
このバックエンドのアーキテクチャとして、オニオンアーキテクチャ、ドメイン駆動設計を採用していて、
依存関係がコントローラー層→ユースケース層
ユースケース層→ドメイン層
インフラ層→ドメイン層（これは依存というより実装）
こんな感じの依存関係を目指しています。
現在、.serena/memoriesにドキュメントをまとめている最中です。
（issue発行するの忘れていたので、先ほど発行させていただきました。）
https://github.com/assari-harassment/voicebooklm-backend/issues/58

現在ののっ君の実装では、このコントローラー層がドメイン層に依存しているので、よくない実装になっています。

junhat6 commented on src/main/kotlin/com/assari/voicebooklm/usecase/memo/GetTagsUseCase.kt:
```diff
@@ -0,0 +1,55 @@
+    val name: String,
+    val count: Int
+)
+
```
tagWithCountはドメインモデルとして管理する方法がいいのかなと考えています。
ユースケース層では単純なinput outputだけを情報として持っておきたいというのもあるんですが、
返すものとしてtagとそのカウントを返すというのは人気のタグという感じで使うためにcountをつけて返すので、その時ってユーザーの体験を意識している。つまりドメインを意識しているっていうことだと思うんですよ。
なので、ドメインモデルとして切り出すのが適切かなと思います。
また、切り出す時にinitのところでバリデーションを書いて、そのドメインモデルが生成されるときに有効なものしか生成されないようにして欲しいです。
（既存のドメインモデル適当に書きすぎて、全然守ってないやそういえば。。。。）

junhat6 reviewed (changes requested):
既存のコードが汚いすぎて困惑していると思いますが、レビュー対応よろしくお願いいたします。

Nox8931 commented on src/main/kotlin/com/assari/voicebooklm/presentation/controller/tag/TagController.kt:
```diff
@@ -0,0 +1,66 @@
+            responseCode = "401",
+            description = "認証失敗",
+            content = [Content(schema = Schema(implementation = ErrorResponse::class))]
+        )
+    )
+    suspend fun getTags(
+        @AuthenticationPrincipal userId: UUID?
+    ): ResponseEntity<TagResponse> {
+        userId ?: return ResponseEntity.status(HttpStatus.UNAUTHORIZED).build()
+
```
nullチェックを維持します。理由は以下の通りです。
理由
防御的プログラミング
JwtAuthenticationFilterの実装変更やエッジケースに備える
認証情報が部分的に設定される可能性を考慮
一貫性
他のコントローラー（MemoController、VoiceController、AuthController）も同様のパターン
コードベース全体の一貫性を維持
実用性
コストが低く、安全性を確保
テストでもこのパターンで動作確認済み

Nox8931 committed:
feat: タグ一覧取得機能をJDBC実装に統一（レビュー対応）

- JPA実装（MemoJpaDataRepository, VoiceMemoRepositoryImpl）を削除
- TagUsageドメインモデルを追加
- TagCountProjection（JDBC用）を追加
- MemoJdbcRepositoryにタグ取得クエリを実装
- VoiceMemoRepositoryImplにタグ取得メソッドを実装
- 不要なドキュメント（MANUAL_TESTING.md）を削除
- Swaggerで動作確認済み

junhat6 commented on src/main/kotlin/com/assari/voicebooklm/config/UseCaseConfig.kt:
このファイルは必要ないかなと思います。UseCaseに@serviceアノテーションを付与すればこのconfigファイル必要なくなるんじゃないかなと思います。
下のgetTagsUseCase.ktにも記載
